### PR TITLE
Use templating for generated files

### PR DESCRIPTION
**Pull request by matiasilva**
This PR adds templating for the various generated files, allowing us to separate logic from text formatting. It pulls out a considerable amount of strings and text into files. This makes the main code a lot cleaner and readable. It also means we can easily extend the project generator to do many more things if needed.

### CODE REVIEW

**Comment by JamesH65** [comment]
That looks much better - nice one. The problem with this being my first Python program is you simply don't know what libraries exist to help you out!

File: templates/fragments.txt [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+gpio_set_dir(PIN_CS, GPIO_OUT);
+gpio_put(PIN_CS, 1);
+{% endmacro %}
+
+{% macro i2c_define() %}
+// I2C defines
+// This example will use I2C0 on GPIO8 (SDA) and GPIO9 (SCL) running at 400KHz.
+// Pins can be changed, see the GPIO function select table in the datasheet for information on GPIO assignments
+#define I2C_PORT i2c0
+#define I2C_SDA 8
```
Comment (lurch) [inline templates/fragments.txt:49]: @kilograham I think this project-generator was created before we added the "board headers" to the Raspberry Pi Pico SDK. Should we now update the pin-numbers used here (for the various interfaces) to match up with https://github.com/raspberrypi/pico-sdk/blob/master/src/boards/include/boards/pico.h ?

File: pico_project.py [hunk 2]
```diff
@@ -823,64 +711,29 @@ def ParseCommandLine():
...
 
     if (features):
-
         # Add any includes
+        includes = []
         for feat in features:
             if (feat in features_list):
-                o = '#include "' +  features_list[feat][H_FILE] + '"\n'
-                file.write(o)
+                includes.append(f'#include "{features_list[feat][H_FILE]}"')
```
Comment (lurch) [inline pico_project.py:722]: Just a personal viewpoint, but if this was me I'd make `includes` just be a list of filenames (which e.g. gives more flexibility for additional processing later), and then in the template do e.g.
```
{% for include in includes %}
include "{{include}}"
{% endfor %}
```
(I'm not familiar with Jinja, so I may not have got the syntax right.)

Comment (JamesH65) [reply pico_project.py:722]: Makes sense.

File: pico_project.py [hunk 3]
```diff
@@ -823,64 +711,29 @@ def ParseCommandLine():
...
-        filename = Path(folder) / (projectName + '.c')
-
-    file = open(filename, 'w')
-
-    main = ('#include <stdio.h>\n'
-            '#include "pico/stdlib.h"\n'
-            )
-    file.write(main)
+    template = jinja_env.get_template('main.txt')
+    mapping = dict(includes="", defines="", initialisers="")
```
Comment (lurch) [inline pico_project.py:715]: I dunno what the various Python style-guides recommend, but personally I'd use:
```
mapping = {'includes': [], 'defines': [], 'initialisers': []}
```

**Comment by matiasilva** [comment]
There's quite a large diff in the main Python file, mostly owing to large blocks of text being removed but also due to applying `autopep8` (https://www.python.org/dev/peps/pep-0008/).

**Comment by JamesH65** [comment]
Might be worth applying autopep8 first on a new branch, to keep whitespace and technical changes separate - makes doing a review diff difficult if they are mixed together.

File: pico_project.py [hunk 4]
```diff
@@ -313,7 +325,8 @@ def __init__(self, parent, initial_config):
     def init_window(self, args):
         self.configure(background=GetBackground())
         self.title("Advanced Configuration")
-        ttk.Label(self, text="Select the advanced options you wish to enable or change. Note that you really should understand the implications of changing these items before using them!").grid(row=0, column=0, columnspan=5)
+        ttk.Label(self, text="Select the advanced options you wish to enable or change. Note that you really should understand the implications of changing these items before using them!").grid(
```
Comment (lurch) [inline pico_project.py:328]: If you're adding line-wrapping to stop the line-length going over some limit, it probably makes sense to build this string by joining several sub-strings, or by using the `\` line-continuation?
E.g. https://stackoverflow.com/questions/4172448/is-it-possible-to-break-a-long-line-to-multiple-lines-in-python/4172465

**Comment by matiasilva** [comment]
> Might be worth applying autopep8 first on a new branch, to keep whitespace and technical changes separate - makes doing a review diff difficult if they are mixed together.

Yes, good idea. See #40.

File: templates/fragments.txt (hunk 1 above)
Comment (JamesH65) [reply templates/fragments.txt:49]: Needs to be a separate issue.

**Comment by JamesH65** [comment]
Going to have to revert this as found a load of problems.

**Pull request merged**